opm/common/utility/CSRGraphFromCoordinates_impl.hpp

```diff
@@ -0,0 +1,546 @@
+    if (nRows < maxNumVertices) {
+        this->ia_.insert(this->ia_.end(),
+                         maxNumVertices - nRows,
+                         this->startPointers().back());
+    }
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::sortColumnIndicesPerRow()
+{
+    // Transposition is, in this context, effectively a linear time (O(nnz))
+    // bucket insertion procedure.  In other words transposing the structure
+    // twice creates a structure with column indices in (ascendingly) sorted
+    // order.
+
+    this->transpose();
+    this->transpose();
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::condenseDuplicates()
+{
+    // Note: Must be called *after* sort().
+
+    const auto colIdx = this->ja_;
+    auto end          = colIdx.begin();
+
+    this->ja_.clear();
+
+    [[maybe_unused]] auto compressedIdx = this->compressedIdx_;
+    if constexpr (TrackCompressedIdx) {
+        this->compressedIdx_.clear();
+    }
+
+    const auto numRows = this->ia_.size() - 1;
+    for (auto row = 0*numRows; row < numRows; ++row) {
+        auto begin = end;
+
+        std::advance(end, this->ia_[row + 1] - this->ia_[row + 0]);
+
+        const auto q = this->ja_.size();
+
+        this->condenseAndTrackUniqueColumnsForSingleRow(begin, end);
+
+        this->ia_[row + 0] = q;
+    }
+
+    if constexpr (TrackCompressedIdx) {
+        this->remapCompressedIndex(std::move(compressedIdx));
+    }
+
+    // Record final table sizes.
+    this->ia_.back() = this->ja_.size();
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::preparePushbackRowGrouping(const int         numRows,
+                                const Neighbours& rowIdx)
+{
+    assert (numRows >= 0);
+
+    this->ia_.assign(numRows + 1, 0);
+
+    for (const auto& row : rowIdx) {
+        this->ia_[row + 1] += 1;
+    }
+
+    // Note index range: 1..numRows inclusive.
+    for (typename Start::size_type i = 1, n = numRows; i <= n; ++i) {
+        this->ia_[0] += this->ia_[i];
+        this->ia_[i]  = this->ia_[0] - this->ia_[i];
+    }
+
+    assert (this->ia_[0] == rowIdx.size());
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::groupAndTrackColumnIndicesByRow(const Neighbours& rowIdx,
+                                     const Neighbours& colIdx)
+{
+    assert (this->ia_[0] == rowIdx.size());
+
+    const auto nnz = rowIdx.size();
+
+    this->ja_.resize(nnz);
+
+    if constexpr (TrackCompressedIdx) {
+        this->compressedIdx_.clear();
+        this->compressedIdx_.reserve(nnz);
+    }
+
+    for (auto nz = 0*nnz; nz < nnz; ++nz) {
+        const auto k = this->ia_[rowIdx[nz] + 1] ++;
```

Comment:
This was a bit tricky, but I think I understand it. If using partial_sum() above, one would need to be careful to make the correct structure for this line to work! (Starting with [0, 0, n0, n1, ...] if n0, n1 are the number of neighbours of the first two nodes.)

Reply:
> f using `partial_sum()` above, one would need to be careful to make the correct structure for this line to work! (Starting with [0, 0, n0, n1, ...] if n0, n1 are the number of neighbours of the first two nodes.)

Yeah, something like that.  I admit I haven't given too much thought to what I'd need to do in order to coax `partial_sum()` into doing the right thing here.